ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java

```diff
@@ -76,34 +78,36 @@ public void addFunction(final KsqlFunction ksqlFunction) {
   }
 
   @Override
-  public UdfFactory ensureFunctionFactory(final UdfFactory factory) {
+  public synchronized UdfFactory ensureFunctionFactory(final UdfFactory factory) {
     validateFunctionName(factory.getName());
 
-    synchronized (lock) {
-      final String functionName = factory.getName().toUpperCase();
-      if (udafs.containsKey(functionName)) {
-        throw new KsqlException("UdfFactory already registered as aggregate: " + functionName);
-      }
-
-      final UdfFactory existing = udfs.putIfAbsent(functionName, factory);
-      if (existing != null && !existing.matches(factory)) {
-        throw new KsqlException("UdfFactory not compatible with existing factory."
-            + " function: " + functionName
-            + " existing: " + existing
-            + ", factory: " + factory);
-      }
-
-      return existing == null ? factory : existing;
+    final String functionName = factory.getName().toUpperCase();
+    if (udafs.containsKey(functionName)) {
```

Comment:
This should check to see if a `udtf` with the same name is already registered.

As per previous discussion [here](https://github.com/confluentinc/ksql/pull/3589#discussion_r336432143), consider moving the code duplication across the three mutating methods to a single `boolean isFunctionRegistered(functionName)`  method, which checks all three collections and us it in `ensureFunctionFactory`, `addAggregateFunctionFactory` and `addTableFunctionFactory`.

Reply:
This doesn't really work in this method as we don't want to check in the udfs map as the method allows re-registering udfs with the same metadata.

If we really want to make this change, I think we can consider this in a further PR https://github.com/confluentinc/ksql/issues/3675